sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java

```diff
@@ -64,53 +69,92 @@ public GeneratedColumns(InputFactory inputFactory,
                             TransactionContext txnCtx,
                             Validation validation,
                             ReferenceResolver<CollectExpression<T, ?>> refResolver,
-                            Collection<Reference> presentColumns,
                             List<GeneratedReference> allGeneratedColumns,
                             List<Reference> allDefaultExpressionColumns) {
-        toValidate = validation == Validation.NONE ? Collections.emptyMap() : new HashMap<>();
+        this.validation = validation;
         InputFactory.Context<CollectExpression<T, ?>> ctx = inputFactory.ctxForRefs(txnCtx, refResolver);
         toInject = new HashMap<>();
+        nestedReferenceByTopLevelColumnName = new HashMap<>();
+
         for (Reference colWithDefault : allDefaultExpressionColumns) {
-            if (!presentColumns.contains(colWithDefault)) {
-                toInject.put(colWithDefault, ctx.add(colWithDefault.defaultExpression()));
-            }
+            toInject.put(colWithDefault, ctx.add(colWithDefault.defaultExpression()));
         }
+
         for (GeneratedReference generatedCol : allGeneratedColumns) {
-            if (presentColumns.contains(generatedCol)) {
-                if (validation == Validation.VALUE_MATCH) {
-                    toValidate.put(generatedCol, ctx.add(generatedCol.generatedExpression()));
-                }
-            } else {
-                toInject.put(generatedCol, ctx.add(generatedCol.generatedExpression()));
-            }
+            processGeneratedExpressionColumn(generatedCol, ctx.add(generatedCol.generatedExpression()));
         }
         expressions = ctx.expressions();
     }
 
+    private void processGeneratedExpressionColumn(Reference reference,
+                                                  Input<?> expressionInput) {
+        toInject.put(reference, expressionInput);
+        final ColumnIdent ident = reference.column();
+        if (!ident.isTopLevel()) {
+            List<Reference> references = nestedReferenceByTopLevelColumnName.getOrDefault(
+                ident.name(),
+                new ArrayList<>());
```

Comment:
I think a manual `.get()` followed by an `if (.. null)` would be more efficient as right now the `new ArrayList<>()` is always evaluated, not just when there is no element found.